### ISSUE TITLE
endorsed-reference-triple

### ISSUE THREAD

**Issue by nedmsmith** [ietf-rats-wg/draft-ietf-rats-corim#2]
To handle a class of endorsements that are conditionally asserted based on acceptance of specific reference value(s), a new triple is needed.
```
;
; Triples map extension
;
 $$triples-map-extension //= (
  comid.endorsed-reference-triples => [+ endorsed-reference-triple-record]
)
comid.endorsed-reference-triples = 7

endorsed-reference-triple-record = [
  environment-map ; target environment
  [ + endorsed-reference-map ] ; reference values with endorsed values
]
endorsed-reference-map = non-empty<{
  ? comid.pos => uint ; position - avoids using mkey
  ? comid.refv => measurement-map, ; reference values
  ? comid.endv => measurement-map / measurement-values-map ; endorsed values
}>
comid.refv = 0
comid.endv = 1
comid.pos = 2
```

**Comment by thomas-fossati** [comment]
refinement proposal:

* remove `pos`: the order in the array matters
* remove optionality of refv and endv

```
$$triples-map-extension //= (
  comid.endorsed-reference-triples => [+ endorsed-reference-triple-record]
)
comid.endorsed-reference-triples = 7

endorsed-reference-triple-record = [
  environment-map ; target environment
  [ + endorsed-reference-map ] ; reference values with endorsed values
]

endorsed-reference-map = [
  comid.refv => measurement-map, ; reference values
  comid.endv => measurement-map / measurement-values-map ; endorsed values
]
```

**Comment by nedmsmith** [comment]
Noting the point that if mkey and pos are removed and position semantics are inferred from the array context which is part of the triple object. Then there is a possibility for Verifier undecidability if two or more tags are created for the same triple subject that have the same type measurements in the triple object. If the measurements are intended NOT as time series data but as set data the Verifier would accept the Evidence if it matched one of the reference values in the set. If the measurements are intended as series data then the Verifier is undecided as to which measurement is first, second etc... as each tag would have a measurement at position 0. 

If 'pos' is retained, then existence of pos addresses both forms of ambiguity. If pos is omitted, then set semantics are applied across all measurements from multiple tags (for the same environment e.g., class-map/class-id). If pos is supplied then there should be a strict ordering 0 to n even though there may be multiple tags supplying measurements. If not (e.g., two measurements at pos = 0) results in an error. 

I don't know if there is a use case for mkey where mkey type isn't time series related. If time series use case is the only motivation for mkey, then I prefer adding 'pos' to reference-endorsed-map/-record vs. mkey.

**Comment by nedmsmith** [comment]
Here is additional use case context for conditional endorsement. A vendor might monitor target environment (TE) vulnerabilities over a period of time and maintain a vulnerabilities database (e.g., https://cve.mitre.org/ ). They might use SVN (or version #) to track changes to the TE where a vulnerability is fixed. When the vendor produces a RV for the TE, it also may provide a list of known vulnerabilities (CVE). Since the verifier needs to obtain current / fresh RVs, it is convenient to have known CVEs included as endorsed values. 

Pragmatically, the CoRIM / CoMID schema using only reference-values-triple and endorsed-values-triple doesn't allow conditional endorsements to be correlated easily. When an evidence / RV match is found, the verifier must perform a more detailed query (one that includes matching on specific measurement values) to know if a given endorsed-value applies. This is like an inner-loop function that many implementation will want to optimize. If the reference-values-triple is pre-loaded with the endorsements for the inner-loop processing the verifier doesn't need to fetch, parse, pre-process and stage endorsed-values for inner-loop optimization. 
For many vendors, they are most aware of which vulnerabilities affect the TE they mfr and know which patches / updates to apply to resolve them. They're motivated to ensure verifiers are equipped with vulnerability and other info needed for an accurate and timely appraisal.

**Comment by nedmsmith** [comment]
The current proposal is therefore:
```
$$triples-map-extension //= (
  comid.reference-endorsed-triples => [ + reference-endorsed-triple-record ]
)
comid.reference-endorsed-triples = 7

reference-endorsed-triple-record = [
  environment-map ; identifies the target environment
  [ + reference-endorsed-map ] ; reference values with endorsed values
]

reference-endorsed-map = {
  &(ref-vals :0) => [ + measurement-map / measurement-values-map ]
  &(end-vals :1) => [ + measurement-map / measurement-values-map ]
}
```

**Comment by nedmsmith** [comment]
Updating proposed CDDL to reflect convention that we disallow multiple instances of the same type measurement.

```
reference-endorsed-triple-record = [
  environment-map ; identifies the target environment
  ref-map ; reference measurements that match evidence
  end-map ; endorsements that are accepted if above reference values match evidence
]
 
ref-map = {
  &(ref-vals :0) =>  measurement-map / measurement-values-map ; no time series data as reference values is possible
}
 
end-map = {
  &(end-vals :0) => [ + measurement-map / measurement-values-map ] ; array denotes time series data – array index 0 is the most recent value, index n is most historical value
}
```
Note: the use of ref-map and end-map may be pedestrian, but helps to illustrate the part handles time series and the part that handles conditional endorsement.

**Comment by nedmsmith** [comment]
resolved by PR #36

**Comment by nedmsmith** [comment]
Resolved by PR #36
